autoprecompiles/src/memory_optimizer.rs

```diff
@@ -230,35 +217,57 @@ impl<T: FieldElement, V: Hash + Eq + Clone + Ord + Display> MemoryAddressCompara
 
         let memory_addresses = addresses
             .map(|addr| {
-                let addr = addr.transform_var_type(&mut |v| v.into());
-                (
-                    addr.clone(),
-                    find_equivalent_expressions(&addr, &constraint_system),
-                )
+                let addr = addr
+                    .into_iter()
+                    .map(|v| v.transform_var_type(&mut |v| v.into()))
+                    .collect::<Vec<_>>();
+                let equivalent_expressions = addr
+                    .iter()
+                    .map(|addr| {
+                        // TODO: Shortcut for constant expressions?
+                        find_equivalent_expressions(addr, &constraint_system)
+                    })
+                    .multi_cartesian_product()
+                    .collect::<Vec<_>>();
+                (addr, equivalent_expressions)
             })
             .collect();
-        Self { memory_addresses }
+        Self {
+            memory_addresses,
+            _marker: PhantomData,
+        }
     }
 
     /// Returns true if we can prove that for two addresses `a` and `b`,
     /// `a - b` cannot be 0.
     pub fn are_addrs_known_to_be_different(
         &self,
-        a: &(T, GroupedExpression<T, V>),
-        b: &(T, GroupedExpression<T, V>),
-        rc: impl RangeConstraintProvider<T, V> + Clone,
+        a: &Address<A>,
+        b: &Address<A>,
+        rc: impl RangeConstraintProvider<A::PowdrField, AlgebraicReference> + Clone,
     ) -> bool {
-        if a.0 != b.0 {
```

Comment:
Address space is now implicitly handled, by relying on the `Eq` implementation of the `Address` type.